packages/ddp-client/common/livedata_connection.js

```diff
@@ -568,6 +634,79 @@ export class Connection {
+          .withValue(invocation, stubInvocation);
+      } catch (e) {
+        stubOptions.exception = e;
+      }
+    }
+    return this._apply(name, stubOptions, args, options, callback);
+  }
+
+  /**
+   * @memberOf Meteor
+   * @importFromPackage meteor
+   * @alias Meteor.applyAsync
+   * @summary Invoke a method passing an array of arguments.
+   * @locus Anywhere
+   * @param {String} name Name of method to invoke
+   * @param {EJSONable[]} args Method arguments
+   * @param {Object} [options]
+   * @param {Boolean} options.wait (Client only) If true, don't send this method until all previous method calls have completed, and don't send any subsequent method calls until this one is completed.
+   * @param {Function} options.onResultReceived (Client only) This callback is invoked with the error or result of the method (just like `asyncCallback`) as soon as the error or result is available. The local cache may not yet reflect the writes performed by the method.
+   * @param {Boolean} options.noRetry (Client only) if true, don't send this method again on reload, simply call the callback an error with the error code 'invocation-failed'.
+   * @param {Boolean} options.throwStubExceptions (Client only) If true, exceptions thrown by method stubs will be thrown instead of logged, and the method will not be invoked on the server.
+   * @param {Boolean} options.returnStubValue (Client only) If true then in cases where we would have otherwise discarded the stub's return value and returned undefined, instead we go ahead and return it. Specifically, this is any time other than when (a) we are already inside a stub or (b) we are in Node and no callback was provided. Currently we require this flag to be explicitly passed to reduce the likelihood that stub return values will be confused with server return values; we may improve this in future.
+   * @param {Function} [asyncCallback] Optional callback.
+   */
+  async applyAsync(name, args, options, callback) {
+    const { stubInvocation, invocation, ...stubOptions } = this._stubCall(name, EJSON.clone(args), options);
+    if (stubOptions.hasStub) {
+      if (
+        !this._getIsSimulation({
+          alreadyInSimulation: stubOptions.alreadyInSimulation,
+          isFromCallAsync: stubOptions.isFromCallAsync,
+        })
+      ) {
+        this._saveOriginals();
+      }
+      try {
+        /*
+         * The code below follows the same logic as the function withValues().
+         *
+         * But as the Meteor package is not compiled by ecmascript, it is unable to use newer syntax in the browser,
+         * such as, the async/await.
+         *
+         * So, to keep supporting old browsers, like IE 11, we're creating the logic one level above.
+         */
+        const currentContext = DDP._CurrentMethodInvocation._setNewContextAndGetCurrent(
+          invocation
+        );
+        try {
+          stubOptions.stubReturnValue = await stubInvocation();
```

Comment:
With the current implementation, it might be difficult to use for packages that create functions to call each method, such as [zodern:relay](https://atmospherejs.com/zodern/relay), [grubba:rpc](https://atmospherejs.com/grubba/rpc), [mdg:validated-method](https://atmospherejs.com/mdg/validated-method). Using `Meteor.call` would be broken for async method stubs, and `Meteor.callAsync` would cause too many problems to use it by default.

I think if the microtask delay here  before calling `DDP._CurrentMethodInvocation._set` is removed  when the stub is sync, it would be fine for these packages to always use Meteor.callAsync. Unless there is something I am missing, when the stub is sync it should then work identically to `Meteor.call`, besides returning a promise.

Something like this might work:
```js
try {
  let stubResult = stubInvocation();
  if (isThenable(stubResult) {
    stubResult = await stubResult;
  }
} finally {
  DDP._CurrentMethodInvocation._set(currentContext);
}
```

Reply:
Implementing this on this [PR](https://github.com/meteor/meteor/pull/12275).